[Indexing] Don't verify mangling of USRs

ahoppen opened the pull request:
Verifying USR mangling adds ~30% overhead to indexing times. Since an incorrect USR mangling doesn't result in a correctness issue at the same level as a miscompile, save those 30% in non-assert builds.

compnerd commented:
```
FAILED: bin/swiftObservation.dll lib/swift/windows/aarch64/swiftObservation.lib 
C:\Windows\system32\cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_dll --intdir=stdlib\public\Observation\Sources\Observation\CMakeFiles\swiftObservation-windows-aarch64.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- T:\5\bin\lld-link.exe /nologo stdlib\public\Observation\Sources\Observation\WINDOWS\aarch64\Observation.obj lib\swift\windows\aarch64\swiftrt.obj stdlib\public\Observation\Sources\Observation\CMakeFiles\swiftObservation-windows-aarch64.dir\ThreadLocal.cpp.obj  /out:bin\swiftObservation.dll /implib:lib\swift\windows\aarch64\swiftObservation.lib /pdb:bin\swiftObservation.pdb /dll /version:0.0 /INCREMENTAL:NO /OPT:REF /OPT:ICF /INCREMENTAL:NO -LIBPATH:T:\aarch64-unknown-windows-msvc\LLVM\.\lib   -LIBPATH:T:\aarch64-unknown-windows-msvc\Runtime\.\lib\swift\windows\aarch64   -LIBPATH:T:\5\bin\..\lib\swift\windows\aarch64   -LIBPATH:T:\5\bin\..\lib\swift\windows   -LIBPATH:\\usr\lib\swift   -LIBPATH:T:\aarch64-unknown-windows-msvc\Runtime\winsdk_lib_aarch64_symlinks lib\swift\windows\aarch64\swift_Concurrency.lib  lib\swift\windows\aarch64\swiftCore.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK: command "T:\5\bin\lld-link.exe /nologo stdlib\public\Observation\Sources\Observation\WINDOWS\aarch64\Observation.obj lib\swift\windows\aarch64\swiftrt.obj stdlib\public\Observation\Sources\Observation\CMakeFiles\swiftObservation-windows-aarch64.dir\ThreadLocal.cpp.obj /out:bin\swiftObservation.dll /implib:lib\swift\windows\aarch64\swiftObservation.lib /pdb:bin\swiftObservation.pdb /dll /version:0.0 /INCREMENTAL:NO /OPT:REF /OPT:ICF /INCREMENTAL:NO -LIBPATH:T:\aarch64-unknown-windows-msvc\LLVM\.\lib -LIBPATH:T:\aarch64-unknown-windows-msvc\Runtime\.\lib\swift\windows\aarch64 -LIBPATH:T:\5\bin\..\lib\swift\windows\aarch64 -LIBPATH:T:\5\bin\..\lib\swift\windows -LIBPATH:\\usr\lib\swift -LIBPATH:T:\aarch64-unknown-windows-msvc\Runtime\winsdk_lib_aarch64_symlinks lib\swift\windows\aarch64\swift_Concurrency.lib lib\swift\windows\aarch64\swiftCore.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST:EMBED,ID=2" failed (exit code 1) with the following output:
lld-link: error: could not open 'swiftWinSDK.lib': no such file or directory

[344/389][ 88%][361.882s] Linking CXX shared library bin\swiftWinSDK.dll
```

Seems like a bug with the Observation module's dependencies (cc: @phausler)

compnerd commented:
@swift-ci please test Windows platform

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[Indexing] Don't verify mangling of USRs

Verifying USR mangling adds ~30% overhead to indexing times. Since an incorrect USR mangling doesn't result in a correctness issue at the same level as a miscompile, save those 30% in non-assert builds.

ahoppen commented on include/swift/AST/ASTMangler.h:
```diff
@@ -391,7 +391,8 @@ class ASTMangler : public Mangler {
   std::string mangleAnyDecl(const ValueDecl *Decl, bool prefix,
-                            bool respectOriginallyDefinedIn = false);
+                            bool respectOriginallyDefinedIn = false,
+                            bool verify = true);
```
Sorry, missed this comment.

Good call with using `prefix`, that really simplifies things. I also just noticed that we were actually verifying USRs more than mangled names during compilation (which are guarded by `#ifndef NDEBUG`).

We can’t sink the check into `finalize` because we don’t know which decl we were mangling in there, so we can’t check for `Decl->isInvalid()`. I just guarded the `verify` call here by an additional `CONDITIONAL_ASSERT_enabled() && !prefix`

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented on lib/AST/ASTMangler.cpp:
```diff
@@ -1026,8 +1032,11 @@ std::string ASTMangler::mangleAccessorEntityAsUSR(AccessorKind kind,
   appendAccessorEntity(getCodeForAccessorKind(kind), decl, isStatic);
   // We have a custom prefix, so finalize() won't verify for us. If we're not
   // in invalid code (coming from an IDE caller) verify manually.
+#ifndef NDEBUG
```
Simplified this a little bit based on Hamish’s comment.
